torchtext/experimental/datasets/raw/translation.py

```diff
@@ -0,0 +1,576 @@
+
+    extracted_files = []
+    if isinstance(URLS[dataset_name], list):
+        for f in URLS[dataset_name]:
+            dataset_tar = download_from_url(f, root=root)
+            extracted_files.extend(extract_archive(dataset_tar))
+    elif isinstance(URLS[dataset_name], str):
+        dataset_tar = download_from_url(URLS[dataset_name], root=root)
+        extracted_files.extend(extract_archive(dataset_tar))
+    else:
+        raise ValueError(
+            "URLS for {} has to be in a form or list or string".format(
+                dataset_name))
+
+    # Clean the xml and tag file in the archives
+    file_archives = []
+    for fname in extracted_files:
+        if 'xml' in fname:
+            _clean_xml_file(fname)
+            file_archives.append(os.path.splitext(fname)[0])
+        elif "tags" in fname:
+            _clean_tags_file(fname)
+            file_archives.append(fname.replace('.tags', ''))
+        else:
+            file_archives.append(fname)
+
+    data_filenames = defaultdict(dict)
+    data_filenames = {
+        "train": _construct_filepaths(file_archives, src_train, tgt_train),
+        "valid": _construct_filepaths(file_archives, src_eval, tgt_eval),
+        "test": _construct_filepaths(file_archives, src_test, tgt_test)
+    }
+
+    for key in data_filenames.keys():
+        if len(data_filenames[key]) == 0 or data_filenames[key] is None:
+            raise FileNotFoundError(
+                "Files are not found for data type {}".format(key))
+
+    datasets = []
+    for key in data_filenames.keys():
+        src_data_iter = _read_text_iterator(data_filenames[key][0])
+        tgt_data_iter = _read_text_iterator(data_filenames[key][1])
+
+        datasets.append(RawTextIterableDataset(src_data_iter, tgt_data_iter))
+
+    return tuple(datasets)
+
+
+class RawTextIterableDataset(torch.utils.data.IterableDataset):
+    """Defines an abstraction for raw text iterable datasets.
+    """
+    def __init__(self, src_iterator, tgt_iterator):
+        """Initiate text-classification dataset.
+        """
+        super(RawTextIterableDataset, self).__init__()
+        self._src_iterator = src_iterator
+        self._tgt_iterator = tgt_iterator
+        self.has_setup = False
+        self.start = 0
+        self.num_lines = None
+
+    def setup_iter(self, start=0, num_lines=None):
+        self.start = start
+        self.num_lines = num_lines
+        self.has_setup = True
+
+    def __iter__(self):
+        if not self.has_setup:
+            self.setup_iter()
+
+        for i, item in enumerate(zip(self._src_iterator, self._tgt_iterator)):
+            if i >= self.start:
+                yield item
+            if (self.num_lines is not None) and (i == (self.start +
+                                                       self.num_lines)):
+                break
+
+    def get_iterator(self):
+        return self._iterator
+
+
+def Multi30k(languages="de-en",
+             train_filename="train",
```

Comment:
Seems to me that we can combine `languages` and `train_filename` together and explicitly the train data as `train_filename`, for example `train_filename="train.de"`. Same for valid_filename.